lib/mongoid/traversable.rb

```diff
@@ -27,7 +27,7 @@ def _parent=(p)
     #
     # @return [ Array<Document> ] All child documents in the hierarchy.
     def _children
-      @__children ||= collect_children
```

Comment:
Yeah, that can work, you just need to make sure you think through all the cases of where this can change (new/create/delete -- does this need to happen for foreign key relations as well or just embedded, etc.) and I imagine clear the memoization all the way up to the root. I'm not familiar enough with the internals to sketch out all the scenarios in which this can happen and feel comfortable that I've addressed them all, though.